Verify user emails

allisonking opened the pull request:
## Issue(s) Resolved
Closes https://github.com/pubpub/platform/issues/1121

## High-level Explanation of PR

Adds a verify email flow.

## Test Plan
I don't think this will work in the preview environment since we need to use emails.

Happy path:
1. In a browser that is not signed in, visit a public form i.e. http://localhost:3000/c/croccroc/public/forms/review/fill
2. You should see the public sign up page. Fill it out 
3. You should get to a page that says you need to verify your email
4. Open up inbucket at localhost:5432/monitor
5. You should have gotten an email with the subject 'Verify your email'. Click this link
6. You should be redirected to the public form, and a toast saying you've verified your email should appear at the bottom.

Other things to test:
* Request another verification email. This should invalidate the first one, and you should get a second email with a new token
  * You can try using the invalidated link and you will see a token invalid message, request another one
* After step 3 above, instead of verifying your email, try to navigate to another part of the app. You should always be sent to the `/verify` page, but the "Forgot password" flow is still accessible

## Screenshots (if applicable)


https://github.com/user-attachments/assets/8a4c51e5-12ef-4d55-afd3-0993a709644f



## Notes

allisonking commented on core/app/(user)/magic-link/route.ts:
```diff
@@ -16,30 +17,8 @@ const redirectToURL = (
-		new URL(`/not-found?from=${encodeURIComponent(redirectTo)}`, env.PUBPUB_URL),
-		opts
-	);
+	const url = createRedirectUrl(redirectTo, opts?.searchParams);
```
Moved this to a helper file since I found I needed it elsewhere (mostly to easily tack on search params). but I think @tefkah 's lil routes lib would be a lot nicer

allisonking commented on core/app/(user)/magic-link/route.ts:
```diff
@@ -129,6 +108,14 @@ export async function GET(req: NextRequest) {
 
 	const { user: tokenUser, authTokenType } = tokenSettled.value;
 
+	if (!tokenUser.isVerified) {
```
by putting this here in the `magic-link` route, we automatically mark users who go through this route as verified since they would've come from their email. this lets us get the requirement "Forgot password also functions as verification, in case where user signs up, never completes verification, and deletes verification email" for free

allisonking commented on core/app/(user)/verify/ResendVerificationButton.tsx:
```diff
@@ -0,0 +1,38 @@
+	};
+
+	return (
+		<SubmitButton
```
very handy button!!

allisonking commented on core/app/layout.tsx:
```diff
@@ -23,7 +24,9 @@ export default function RootLayout({ children }: { children: React.ReactNode })
 					<NuqsAdapter>
 						<TooltipProvider>
 							{children}
-							<Toaster />
+							<Suspense>
+								<RootToaster />
+							</Suspense>
```
hopefully this is okay—I wasn't sure how else to get the toast after redirect to work. it needs `Suspense` since `RootToaster` needs to `useSearchParams` in order to figure out whether to render the 'verified' toast

allisonking commented on core/app/page.tsx:
```diff
@@ -1,23 +1,36 @@
 	}
 
-	redirect(`/c/${communitySlug}/stages`);
+	redirect(createRedirectUrl(`/c/${communitySlug}/stages`, params).toString());
```
the regular pages weren't passing on search params. I needed it to so that `?verified=true` could get passed along to arbitrary pages

tefkah commented on core/prisma/seed/seedCommunity.ts:
```diff
@@ -95,6 +95,10 @@ export type UsersInitializer = Record<
 			slug?: string;
 			existing?: false;
+			/**
+			 * @default true
```
good default

tefkah commented on core/lib/server/email.tsx:
```diff
@@ -113,6 +111,36 @@ export function passwordReset(
+	return buildSend(async () => {
+		const magicLink = await createMagicLink(
+			{
+				type: AuthTokenType.generic,
```
could you maybe add a small comment here why this needs to be `AuthTokenType.generic` vs `AuthTokenType.verifyEmail`? I think it's bc `/magic-link` will set their email to be verified, but i think that's easy to forget!

tefkah reviewed (approved):
i think it looks great! very elegant solution of making magic-links set the verified to `true`.

I made some optional suggestions which don't really need to be implemented if you don't feel like it, but there's one situation we do still need to cover.

if a user is invited through the email action, they are prompted to fill in their information+create a password. they can pick a different email during signup. i think in that case, we should force them to verify their email. this happens here:
https://github.com/pubpub/platform/blob/a54eb9b77fd202346153a294e3775c9dc9eb9bfb/core/lib/authentication/actions.ts#L493-L529

i think we'll get rid of this route as a special case soon, but we should still cover it for now!

greate work!

allisonking commented:
Thanks @tefkah , did not realize the legacy signup needed this too! added, and also implemented using nuqs, great idea!


https://github.com/user-attachments/assets/8b92e659-888c-4f7c-8274-6acf2036a243
